src/types/transaction.zig

```diff
@@ -68,6 +71,132 @@ pub const Transaction = struct {
+            try w.writeInt(u32, input.previous_outpoint.index, .little);
+            try compact_script_len.encodeToWriter(w);
+            try w.writeAll(input.script_sig.bytes);
+            try w.writeInt(u32, input.sequence, .little);
+        }
+
+        try compact_output_len.encodeToWriter(w);
+        for (self.outputs.items) |output| {
+            const compact_script_len = CompactSizeUint.new(output.script_pubkey.bytes.len);
+
+            try w.writeInt(i64, output.value, .little);
+            try compact_script_len.encodeToWriter(w);
+            try w.writeAll(output.script_pubkey.bytes);
+        }
+
+        try w.writeInt(u32, self.lock_time, .little);
+    }
+
+    /// Serialize a message as bytes and write them to the buffer.
+    ///
+    /// buffer.len must be >= than self.hintSerializedLen()
+    pub fn serializeToSlice(self: *const Self, buffer: []u8) !void {
+        var fbs = std.io.fixedBufferStream(buffer);
+        try self.serializeToWriter(fbs.writer());
+    }
+
+    /// Serialize a message as bytes and return them.
+    pub fn serialize(self: *const Self, allocator: std.mem.Allocator) ![]u8 {
+        const serialized_len = self.virtual_size();
+
+        const ret = try allocator.alloc(u8, serialized_len);
+        errdefer allocator.free(ret);
+
+        try self.serializeToSlice(ret);
+
+        return ret;
+    }
+
+    pub fn deserializeReader(allocator: std.mem.Allocator, r: anytype) !Self {
+        comptime {
+            if (!std.meta.hasFn(@TypeOf(r), "readInt")) @compileError("Expects r to have fn 'readInt'.");
+            if (!std.meta.hasFn(@TypeOf(r), "readNoEof")) @compileError("Expects r to have fn 'readNoEof'.");
+            if (!std.meta.hasFn(@TypeOf(r), "readAll")) @compileError("Expects r to have fn 'readAll'.");
+            if (!std.meta.hasFn(@TypeOf(r), "readByte")) @compileError("Expects r to have fn 'readByte'.");
+        }
+
+        var tx: Self = undefined;
+
+        tx.version = try r.readInt(i32, .little);
+
+        const compact_input_len = try CompactSizeUint.decodeReader(r);
+        var inputs = try std.ArrayList(Input).initCapacity(allocator, compact_input_len.value());
+        errdefer inputs.deinit();
+
+        while (inputs.items.len < compact_input_len.value()) {
+            var input: Input = undefined;
+            var hash_bytes: [32]u8 = undefined;
+            const hash_raw_bytes = try read_bytes_exact(allocator, r, 32);
+            defer allocator.free(hash_raw_bytes);
+            @memcpy(&hash_bytes, hash_raw_bytes);
+
+            input.previous_outpoint.hash = Hash{ .bytes = hash_bytes };
+            input.previous_outpoint.index = try r.readInt(u32, .little);
+            const compact_script_len = (try CompactSizeUint.decodeReader(r)).value();
+
+            input.script_sig = Script{ .bytes = try read_bytes_exact(allocator, r, compact_script_len), .allocator = allocator };
+            input.sequence = try r.readInt(u32, .little);
```

Comment:
Can you do a serialize/deserialize impl for Input and Output directly?

Reply:
Sure. Should I create new files, are transaction.zig is getting pretty big ?

Reply:
yup, you can do one file per struct. And you can use this Zig quick, where the file IS the struct, so no need for `const Input = struct`

https://zig.news/gowind/zig-files-are-structs-288j

Reply:
Same for TransactionHash and so on